src/client/Client.cc

```diff
@@ -9018,8 +9018,7 @@ int Client::_write(Fh *f, int64_t offset, uint64_t size, const char *buf,
     return r;
 
   /* clear the setuid/setgid bits, if any */
-  if (unlikely((in->mode & S_ISUID) ||
-	       (in->mode & (S_ISGID | S_IXGRP)) == (S_ISGID | S_IXGRP))) {
+  if (unlikely(in->mode & (S_ISUID|S_ISGID)) && size > 0) {
```

Comment:
I'm not sure about this. We generally leave setgid bit set if group execute bit is not set. This combination makes no sense, so it was used as an indicator of mandatory locking. FWIW, POSIX omits this from the spec, but it's historical Linux and Unix behavior. It's probably harmless to leave it set in that case and may keep you from running afoul of some other testcases. I don't feel too strongly either way though.